### PR TITLE
IDE Support for Required Members

### PR DESCRIPTION
**Pull request by 333fred**
Initial support for `required` in the IDE. I've implemented and tested completion, property/field generation features (such as generate overrides), and symbol display. An open question I have: should we displaying `required` in QuickInfo? We don't display any modifiers there today, but maybe `required` is special here?

Spec: https://github.com/dotnet/csharplang/blob/main/proposals/required-members.md
Test plan: https://github.com/dotnet/roslyn/issues/57046.

### CODE REVIEW

**Comment by 333fred** [comment]
@jasonmalinowski for the quickinfo question.

**Comment by CyrusNajmabadi** [comment]
> should we displaying required in QuickInfo? We don't display any modifiers there today, but maybe required is special here?

i can't really come up with a good justification that wouldn't also apply to all other modifiers.

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:842]: :D

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateMethodSymbol.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceDelegateMethodSymbol.cs:455]:
```suggestion
```

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs:647]: we're unfortunately inconsistent, but preference is DoNot, not Dont in a member name.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:32]: Consider renaming `AddFieldModifiersIfRequired` too for consistency

**Review by Youssef1313 — COMMENTED** [review]
Consider updating https://github.com/dotnet/roslyn/blob/9d669eff91debe512e5ccb760c4efa73394d8bb9/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/CodeStyle/CSharpIdeCodeStyleOptions.cs#L20-L29 (this is for [IDE0036](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0036). A doc update will be needed after that)

**Review by Youssef1313 — COMMENTED** [review]
Consider testing [Generate overrides](https://github.com/dotnet/roslyn/blob/6b0d35d6dce5787b24b74baae893b85e5cffb191/src/EditorFeatures/CSharpTest/GenerateOverrides/GenerateOverridesTests.cs) and [implement abstract class](https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/CSharpTest/ImplementAbstractClass/ImplementAbstractClassTests.cs) for:

```csharp
using System;
public abstract class C
{
    public virtual required int P1 { get; set; }
    public abstract required int P2 { get; set; }
}

public class C2 : C
{
}
```

**Review by Youssef1313 — COMMENTED** [review]
Consider supporting `required` in [F1 help service](https://github.com/dotnet/roslyn/blob/9d669eff91debe512e5ccb760c4efa73394d8bb9/src/VisualStudio/CSharp/Test/F1Help/F1HelpTests.cs).

**Review by Youssef1313 — COMMENTED** [review]
Consider extending [this test](https://github.com/dotnet/roslyn/blob/9d669eff91debe512e5ccb760c4efa73394d8bb9/src/EditorFeatures/CSharpTest/ExtractInterface/ExtractInterfaceTests.cs#L618-L645) with a required property. Generated interface code shouldn't have `required` modifier, I think.

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:127]: Is https://github.com/dotnet/roslyn/blob/9d669eff91debe512e5ccb760c4efa73394d8bb9/src/Workspaces/Core/Portable/Shared/Extensions/ISymbolExtensions.cs#L28-L37 outdated?

(Not only for `IsRequired`)

Can we figure a case where it's problematic?

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:32]: I would say methods in this area suffixed with 'IfRequired' generally :)

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs
Comment (RikkiGibson) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:1663]: Do any .NET design guidelines need to be updated to recommend a placement of the 'required' modifier?

**Review by RikkiGibson — APPROVED** [review]

**Comment by 333fred** [comment]
> Consider supporting `required` in [F1 help service](https://github.com/dotnet/roslyn/blob/9d669eff91debe512e5ccb760c4efa73394d8bb9/src/VisualStudio/CSharp/Test/F1Help/F1HelpTests.cs).

This can wait for after merge.

> Consider updating
> 
> https://github.com/dotnet/roslyn/blob/9d669eff91debe512e5ccb760c4efa73394d8bb9/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/CodeStyle/CSharpIdeCodeStyleOptions.cs#L20-L29
> 
> (this is for [IDE0036](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0036). A doc update will be needed after that)

This (I think) can wait for after merge. Up to @jasonmalinowski.

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs
Comment (333fred) [reply src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:127]: While I can certainly update this codepath, I'm not immediately seeing a way to observe the update.

Comment (Youssef1313) [reply src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:127]: @333fred Probably generate overrides can observe it? https://github.com/dotnet/roslyn/blob/00405d11948c7ca2c44b1ab276c4cf977b30a06d/src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs#L496-L500

File: src/Compilers/Core/Portable/Symbols/IPropertySymbol.cs
Comment (jcouv) [inline src/Compilers/Core/Portable/Symbols/IPropertySymbol.cs:37]: Do we have compiler tests for added public APIs?

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:32]: I'm curious about this too. Why rename two of the three APIs here?

**Review by jcouv — COMMENTED** [review]
Done with review pass of compiler change (iteration 6)

File: src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs
Comment (333fred) [reply src/Workspaces/Core/Portable/Editing/DeclarationModifiers.cs:127]: AFAICT, it did not, as I didn't update that codepath before writing that test.

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:32]: Because I derped.

File: src/Compilers/Core/Portable/Symbols/IPropertySymbol.cs
Comment (jcouv) [reply src/Compilers/Core/Portable/Symbols/IPropertySymbol.cs:37]: Per offline discussion, will be added in subsequent PR

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7)

**Comment by jasonmalinowski** [comment]
> @CyrusNajmabadi: i can't really come up with a good justification that wouldn't also apply to all other modifiers.

I'm wondering if we _originally_ didn't show modifiers in quick info because a common use of quick info might be to know the type of something; by sticking all the modifiers in front you have to scan past them before you can find the type. I can also think of a lot of scenarios where modifiers would be good though too. So I'm OK with showing all modifiers, but I'll add an entry into the design meeting queue for the IDE to confirm that decision. I wouldn't block this PR on that, and if people don't like it we can easily change it.

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:127]: Double blank line.

File: src/Compilers/Core/Portable/Symbols/IFieldSymbol.cs
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/Symbols/IFieldSymbol.cs:45]:
```suggestion
        /// True if this field is required to be set in an object initializer during construction.
```

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs:2774]: Is this behavior consistent with other modifiers like this? It almost feels like override completion should instead be filtering to required members in that case, i.e. I wrote "required" why is it being ignored and erased?

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.CSharp.cs
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.CSharp.cs:763]: Don't we have a file somewhere where we put constants for stuff like this? I thinki we might have for span/index?

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.CSharp.cs
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.CSharp.cs:782]: Do we have a tracking bug for filtering these out somewhere?

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/RequiredKeywordRecommender.cs
Comment (jasonmalinowski) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/RequiredKeywordRecommender.cs:16]: Should this also include readonly and const?

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs
Comment (jasonmalinowski) [inline src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs:3639]: This isn't valid right? The required would actually need a setter of some kind?

File: src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs
Comment (jasonmalinowski) [inline src/Workspaces/Core/Portable/Shared/Extensions/SyntaxGeneratorExtensions.cs:512]: Blank line.

**Review by jasonmalinowski — DISMISSED** [review]
Some small comments, but nothing I'd hold this PR up for.

**Comment by jasonmalinowski** [comment]
@333fred For the F1 support and modifier reordering, I'd say OK to not do it in this PR, but do it before we merge back to main. Should be small fixes regardless.

**Comment by 333fred** [comment]
> but do it before we merge back to main

This is the pr for that work. When you sign off on this, it should mean that we're good to merge back to main.

**Comment by jcouv** [comment]
> @333fred For the F1 support and modifier reordering, I'd say OK to not do it in this PR, but do it before we merge back to main. Should be small fixes regardless.

In my view, both of those are above the minbar for merging to `main`. But if easy to fix now...

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs
Comment (333fred) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/OverrideCompletionProviderTests.cs:2774]: I don't necessarily think so. Maybe we could make it filter out things that _can't_ be required, but I might want the override of a non-required property to be required by the override.

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.CSharp.cs
Comment (333fred) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.CSharp.cs:782]: https://github.com/dotnet/roslyn/issues/61511

File: src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.CSharp.cs
Comment (333fred) [reply src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.CSharp.cs:763]: Not that I can find.

File: src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs
Comment (333fred) [reply src/Workspaces/CSharpTest/CodeGeneration/SyntaxGeneratorTests.cs:3639]: Correct.

**Review by jasonmalinowski — APPROVED** [review]
:shipit:

**Pull request merged**
